judge/runguard.c

```diff
@@ -1239,72 +1311,38 @@ int main(int argc, char **argv)
 				}
 			}
 
+			errno = 0;
+			verbose("calling pselect, nfds = %d", nfds);
 			r = pselect(nfds+1, &readfds, NULL, NULL, NULL, &emptymask);
+			verbose("pselect done, r = %d, errno = %d = %s", r, errno, strerror(errno));
+
 			if ( r==-1 && errno!=EINTR ) error(errno,"waiting for child data");
 
 			if ( received_SIGCHLD ) {
 				if ( (pid = wait(&status))<0 ) error(errno,"waiting on child");
+				verbose("received_SIGCHLD for pid = %d", pid);
 				if ( pid==child_pid ) break;
 			}
 
-			/* Check to see if data is available and pass it on */
-			for(i=1; i<=2; i++) {
-				if ( child_pipefd[i][PIPE_OUT] != -1 &&
-				     FD_ISSET(child_pipefd[i][PIPE_OUT],&readfds) ) {
-
-					if (limit_streamsize && data_passed[i] == streamsize) {
-						/* Throw away data if we're at the output limit, but
-						   still count how much data we consumed  */
-						nread = read(child_pipefd[i][PIPE_OUT], buf, BUF_SIZE);
-					} else {
-						/* Otherwise copy the output to a file */
-						to_read = BUF_SIZE;
-						if (limit_streamsize) {
-							to_read = min(BUF_SIZE, streamsize-data_passed[i]);
-						}
-
-						if ( use_splice ) {
-							nread = splice(child_pipefd[i][PIPE_OUT], NULL,
-							               child_redirfd[i], NULL,
-							               to_read, SPLICE_F_MOVE);
-
-							if ( nread==-1 && errno==EINVAL ) {
-								use_splice = 0;
-								verbose("splice failed, switching to read/write");
-								/* Setting errno here to repeat the copy. */
-								errno = EAGAIN;
-							}
-						} else {
-							nread = read(child_pipefd[i][PIPE_OUT], buf, to_read);
-							if ( nread>0 ) {
-								nread = write(child_redirfd[i], buf, nread);
-							}
-						}
-
-						if ( nread>0 ) data_passed[i] += nread;
-
-						/* print message if we're at the streamsize limit */
-						if (limit_streamsize && data_passed[i] == streamsize) {
-							verbose("child fd %i limit reached",i);
-						}
-					}
-					if ( nread==-1 ) {
-						if (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK) continue;
-						error(errno,"copying data fd %d",i);
-					}
-					if ( nread==0 ) {
-						/* EOF detected: close fd and indicate this with -1 */
-						if ( close(child_pipefd[i][PIPE_OUT])!=0 ) {
-							error(errno,"closing pipe for fd %d",i);
-						}
-						child_pipefd[i][PIPE_OUT] = -1;
-						continue;
-					}
-					data_read[i] += nread;
-				}
+			pump_pipes(readfds, data_read, data_passed);
+		}
+
+		FD_ZERO(&readfds);
+		nfds = -1;
+		for(i=1; i<=2; i++) {
+			if ( child_pipefd[i][PIPE_OUT]>=0 ) {
+				FD_SET(child_pipefd[i][PIPE_OUT],&readfds);
+				nfds = max(nfds,child_pipefd[i][PIPE_OUT]);
 			}
 		}
 
+		errno = 0;
+		verbose("final call pselect, nfds = %d", nfds);
+		r = pselect(nfds+1, &readfds, NULL, NULL, &pipedelay, &emptymask);
```

Comment:
I haven't tried but when rereading the code, I thought it would be possible to *know* when to read further, by remembering on which pipes we reached EOF. That would make this pselect call unnecessary.

Am I mistaken there?

Reply:
I would assume that even if data is not available right away, the read/ splice is blocking. So we could read until eof.